Nama  :    
Batch :   
Objective :    


# Data Science Salaries

## Introduction

Melakukan analisa terhadap Data Science Salaries melalui data processing. 
Tujuannya untuk belajar.

In [1]:
import pandas as pd

## Data Loading

In [35]:
# Membaca file csv 
data = pd.read_csv("./ds_salaries.csv")
data.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')


11

## Data Knowledge

In [41]:
# Summary data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [42]:
# List column
print(data.columns)
# Total column
len(data.columns)

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')


11

Dataset data science salaries memiliki total 3755 rows dengan 11 column.
11 columns terdiri dari : 
- 'work_year'
- 'experience_level'
- 'employment_type'
- 'job_title','salary'
- 'salary_currency'
-'salary_in_usd'
- 'employee_residence',
- 'remote_ratio'
- 'company_location'
- 'company_size'

In [46]:
dataResidences = data['employee_residence'].unique()
# Total data data residences
print(len(dataResidences))
dataResidences 

78


array(['ES', 'US', 'CA', 'DE', 'GB', 'NG', 'IN', 'HK', 'PT', 'NL', 'CH',
       'CF', 'FR', 'AU', 'FI', 'UA', 'IE', 'IL', 'GH', 'AT', 'CO', 'SG',
       'SE', 'SI', 'MX', 'UZ', 'BR', 'TH', 'HR', 'PL', 'KW', 'VN', 'CY',
       'AR', 'AM', 'BA', 'KE', 'GR', 'MK', 'LV', 'RO', 'PK', 'IT', 'MA',
       'LT', 'BE', 'AS', 'IR', 'HU', 'SK', 'CN', 'CZ', 'CR', 'TR', 'CL',
       'PR', 'DK', 'BO', 'PH', 'DO', 'EG', 'ID', 'AE', 'MY', 'JP', 'EE',
       'HN', 'TN', 'RU', 'DZ', 'IQ', 'BG', 'JE', 'RS', 'NZ', 'MD', 'LU',
       'MT'], dtype=object)

In [48]:
dataCompanySizes = data['company_size'].unique()
# Total data comapny sizes
print(len(dataCompanySizes))
dataCompanySizes

3


array(['L', 'S', 'M'], dtype=object)

Terdapat uniq data pada company size yaitu L, S dan M. Sementara untuk residence terdapat 78 dari seluruh negara.

## Data Cleaning

In [50]:
data.isnull().any()

work_year             False
experience_level      False
employment_type       False
job_title             False
salary                False
salary_currency       False
salary_in_usd         False
employee_residence    False
remote_ratio          False
company_location      False
company_size          False
dtype: bool

Telah dilakukan pengecekan missing value. Tidak ditemukan data yang missing maka tidak ada process data cleaning 

https://blog.hubspot.com/website/duplicated-pandas

In [51]:
data.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [94]:
# duplicate dataframe 
df = data.copy()
# drop 3 column
df = df.drop(['salary','salary_currency','remote_ratio'], axis=1)

df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary_in_usd employee_residence company_location company_size  
0          85847                 ES               ES            L  
1          30000                 US               US            S  
2          25500                 US               US            S  
3         175000                 CA               CA            M  
4         120000                 CA               CA            M

In [95]:
df = df.rename(columns={"employment_type": "type", "salary_in_usd": "salary"})
df.head()

work_year experience_level type                 job_title  salary  \
0       2023               SE   FT  Principal Data Scientist   85847   
1       2023               MI   CT               ML Engineer   30000   
2       2023               MI   CT               ML Engineer   25500   
3       2023               SE   FT            Data Scientist  175000   
4       2023               SE   FT            Data Scientist  120000   

  employee_residence company_location company_size  
0                 ES               ES            L  
1                 US               US            S  
2                 US               US            S  
3                 CA               CA            M  
4                 CA               CA            M

In [96]:
# salary remove currency and symbol 
df['salary'] = df['salary'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   type                3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   employee_residence  3755 non-null   object
 6   company_location    3755 non-null   object
 7   company_size        3755 non-null   object
dtypes: int64(2), object(6)
memory usage: 234.8+ KB


In [97]:
df.head()

work_year experience_level type                 job_title  salary  \
0       2023               SE   FT  Principal Data Scientist   85847   
1       2023               MI   CT               ML Engineer   30000   
2       2023               MI   CT               ML Engineer   25500   
3       2023               SE   FT            Data Scientist  175000   
4       2023               SE   FT            Data Scientist  120000   

  employee_residence company_location company_size  
0                 ES               ES            L  
1                 US               US            S  
2                 US               US            S  
3                 CA               CA            M  
4                 CA               CA            M

In [59]:
df.to_csv('data_clean.csv', index=False)

## Normalisasi

Proses normalisasi hingga 3NF menghasilkan 3 table yaitu table utama merupakan salaries, dan 2 table lainnya adalah residences dan companySizes. Berikut hubungan antar table: 
- 1 Residence memiliki banyak Salary. Melalui primary key id di Residences menjadi foreign di Salaries dengan nama residence_id. 
- 1 CompanySize memiliki banyak Salary. Melalui primary key id di CompanySizes menjadi foreign di Salaries dengan nama companysize_id. 

In [62]:
# membuat csv untuk table residences
dataPandaResidence = pd.DataFrame({
  "residence" : dataResidences
})
dataPandaResidence.to_csv("residences.csv", index=False)

In [65]:
# membuat csv untuk table companySize
dataPandaCompanySize = pd.DataFrame({
  "size":dataCompanySizes
})
dataPandaCompanySize.to_csv("companySizes.csv", index=False)

In [99]:
# print(df["employee_residence"])
# print(dataPandaResidence)
              #  pembanding, list dari csv, new name di main table
def generateFK(name_replace, dataUniq, nameFK):
  result = []
  datalist = dataUniq.to_list()
  # loop data 3000an sekian
  for el in df[name_replace]:
    # list dengan string/calue 
    result.append(datalist.index(el) + 1)
  df[nameFK] = result
  return result


print(generateFK("employee_residence", dataPandaResidence['residence'], 'residence_id'))
print(generateFK("company_size", dataPandaCompanySize['size'], 'companySize_id'))

df = df.drop(['employee_residence','company_size'], axis=1)
df.head()

[1, 2, 2, 3, 3, 2, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 5, 5, 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 6, 6, 7, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 8, 9, 7, 7, 10, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 11, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 7, 5, 5, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 12, 12, 2, 2, 2, 2, 5, 5, 6, 5, 5, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 13, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 2, 7, 14, 15, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 16, 2, 2, 3, 2, 2, 2

work_year experience_level type                 job_title  salary  \
0       2023               SE   FT  Principal Data Scientist   85847   
1       2023               MI   CT               ML Engineer   30000   
2       2023               MI   CT               ML Engineer   25500   
3       2023               SE   FT            Data Scientist  175000   
4       2023               SE   FT            Data Scientist  120000   

  company_location  residence_id  companySize_id  
0               ES             1               1  
1               US             2               2  
2               US             2               2  
3               CA             3               3  
4               CA             3               3

In [100]:
# create main table csv (FIXED)
# df.to_csv("C:\tmp\salaries_table.csv", index=False)
df.to_csv("salaries_table.csv", index=False)

## Relational Database & SQL

In [ ]:
# ikut csv mu untuk nama column.... 

CREATE TABLE residences (
  id SERIAL PRIMARY KEY, 
  residence VARCHAR 
)

CREATE TABLE "companySizes" (
  id SERIAL PRIMARY KEY, 
  size VARCHAR 
)

CREATE TABLE salaries (
  id SERIAL PRIMARY KEY, 
  work_year INT,
  experience_level VARCHAR,
  type VARCHAR,
  job_title VARCHAR,
  salary INT,
  company_location VARCHAR,
  residence_id INT,
  "companySize_id" INT, 
  FOREIGN KEY (residence_id) REFERENCES residences(id),
  FOREIGN KEY ("companySize_id") REFERENCES "companySizes"(id)
)

---

### Info 
1. Business Understanding 
- SMART Analysis
- Problem Statement from SMART Analysis 
- 5W1H 
2. Data Loading
3. Data Knowledge
4. Data Cleaning
5. Data Transform
6. Descriptive Statistic
7. Statistic Inferensial by answer 5W1H